aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,610 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <tuple>
+
+namespace at {
+namespace native {
+
+template <typename scalar_t>
+Tensor max_unpooling2d_forward_out_cpu_frame(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t oheight,
+    int64_t owidth) {
+  int64_t numBatch = 1;
+  int64_t dimc = 0;
+  int64_t dimh = 1;
+  int64_t dimw = 2;
+  if (input.ndimension() == 4) {
+    numBatch = input.size(0);
+    dimc++;
+    dimh++;
+    dimw++;
+  }
+  int64_t numChannels = input.size(dimc);
+  int64_t inputHeight = input.size(dimh);
+  int64_t inputWidth = input.size(dimw);
+
+  auto* rawInput = input.data<scalar_t>();
+  auto* rawIndices = indices.data<int64_t>();
+  auto* rawOutput = output.data<scalar_t>();
+
+  for (int64_t n = 0; n < numBatch; n++) {
+    int64_t nOutputOffset = n * numChannels * owidth * oheight;
+    int64_t nInputOffset = n * numChannels * inputWidth * inputHeight;
+    int64_t k;
+    bool has_error = false;
+    int64_t error_index = 0;
+#pragma omp parallel for private(k)
```

Comment:
These were converted into `at::parallel_for` in the source code (take a look); we should use that function here.

Reply:
Relevant lines:

```
  std::mutex mutex;
  at::parallel_for(0, nslices, 0, [&](int64_t start, int64_t end) {
    for (auto k = start; k < end; k++)
    {
```

and then

```
          if(maxp<0 || maxp>=owidth*oheight) {
            std::unique_lock<std::mutex> lock(mutex);
            has_error = 1;
            error_index = maxp;
```

Reply:
You can also seek guidance from #20032